File: 36-valid-sudoku/valid-sudoku.cs
public class Solution {
    public bool IsValidSudoku(char[][] board) {
        // Initialize hash sets for rows, columns, and sub-boxes
        HashSet<char>[] rows = new HashSet<char>[9];
        HashSet<char>[] cols = new HashSet<char>[9];
        HashSet<char>[] boxes = new HashSet<char>[9];
        
        for (int i = 0; i < 9; i++) {
            rows[i] = new HashSet<char>();
            cols[i] = new HashSet<char>();
            boxes[i] = new HashSet<char>();
        }
        
        // Traverse the board
        for (int i = 0; i < 9; i++) {
            for (int j = 0; j < 9; j++) {
                char current = board[i][j];
                
                // If the cell is empty, continue
                if (current == '.') {
                    continue;
                }
                
                // Check the row
                if (rows[i].Contains(current)) {
                    return false;
                } else {
                    rows[i].Add(current);
                }
                
                // Check the column
                if (cols[j].Contains(current)) {
                    return false;
                } else {
                    cols[j].Add(current);
                }
                
                // Check the 3x3 sub-box
                int boxIndex = (i / 3) * 3 + (j / 3);
                if (boxes[boxIndex].Contains(current)) {
                    return false;
                } else {
                    boxes[boxIndex].Add(current);
                }
            }
        }
        
        return true;
    }
}

File: 36-valid-sudoku/valid-sudoku.js
/**
 * @param {character[][]} board
 * @return {boolean}
 */
var isValidSudoku = function(board) {
    let rows=new Array(9).fill().map(()=>new Set());
    let cols=new Array(9).fill().map(()=> new Set());
    let boxes =new Array(9).fill().map(()=>new Set());

    for (let r=0; r<9; r++){
        for (let c =0; c<9; c++){
            let val=board[r][c];
            if( val==='.') continue;

            let boxIndex=Math.floor(r/3)*3+Math.floor(c/3);

            if(rows[r].has(val) || cols[c].has(val) || boxes[boxIndex].has(val)){
                return false;
            }
            rows[r].add(val);
            cols[c].add(val);
            boxes[boxIndex].add(val);
        }
    }return true;
};